rand_os/src/solarish.rs

```diff
@@ -110,18 +112,48 @@ impl OsRngImpl for OsRng {
     }
 }
 
-fn getrandom(buf: &mut [u8], blocking: bool) -> libc::c_long {
-    extern "C" {
-        fn syscall(number: libc::c_long, ...) -> libc::c_long;
+#[cfg(target_os = "illumos")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::ssize_t;
+#[cfg(target_os = "solaris")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::c_int;
```

Comment:
I stumbled upon this while investigating https://github.com/rust-lang/miri/issues/3924. In this code, the return type is different between Illumos and Solaris. However, in today's libc, the return type is the same for both OSes. (This was added in https://github.com/rust-lang/libc/pull/2236.)

Do you remember where you got the different return type from?

Reply:
It was like this since getrandom v0.1 days: https://docs.rs/crate/getrandom/0.1.0/source/src/solaris_illumos.rs#28-31

We probably just copied what `std` done at the time.

Reply:
In the current version of `getrandom` we [use](https://docs.rs/crate/getrandom/0.2.15/source/src/solaris.rs#23) `libc::getrandom`, so it should no longer be an issue with up-to-date dependencies.

Reply:
Ah getrandom is older than this? Okay.

>  so it should no longer be an issue with up-to-date dependencies.

Yeah, I am just trying to figure out where the other return type comes from and which one is correct...

Reply:
`rand_os` is an older crate. It was eventually deprecated in favor of `getrandom`.

Reply:
I could not find code like this in the history of `std`.

Reply:
If you want to peruse history, the old (pre-rebase) rand history is [here](https://github.com/dhardy/rand/tree/old-master).

I don't remember the details here, but prior to `getrandom` being public API we had `OsRng`.

Reply:
That old branch does not have Solaris at all yet, it seems.

Reply:
I'm not aware of the history behind the original Solaris interface in the crate.  I'm guessing though the difference is that the original version was issuing a direct syscall (since changed to use libc) while the illumos version uses the libc function (which wraps the syscall into something a bit more ergonomic for callers).

As a side note, using the libc function (on both platforms) is the better choice. Unlike Linux, the syscall interface on Solaris and illumos is explicitly not a stable interface and can change incompatibly at any time. Stable interfaces are provided by the system-provided C libraries (and are documented along with an indication of the situations where an interface could change in a possibly incompatible manner -- although even there effort is generally made to avoid breaking consumers even if it's 'allowed').

Reply:
This PR uses dlsym to lookup the `getrandom` symbol. That doesn't look like a direct syscall to me. And yet it uses a different return type for Solaris and Illumos. In the current version of the libc crate, the return type for the Solaris `getrandom` function has changed to `ssize_t`. So either this old PR is wrong, or current libc is wrong, or Solaris changed the signature of their libc `getrandom` function, it seems.

Reply:
Alternatively, this [blog post](https://blogs.oracle.com/solaris/post/solaris-new-system-calls-getentropy2-and-getrandom2) could be wrong and this PR relied on it to get `getrandom` signature since Solaris does not have proper openly published docs.

Reply:
There are other traces pointing at Solaris having a different return type, see https://github.com/rust-lang/miri/issues/3924:

- https://www.gnu.org/software/gnulib/manual/html_node/getrandom.html
- https://docs.oracle.com/cd/E86824_01/html/E54765/getrandom-2.html

Reply:
I mean the version pre-change.

The `getrandom()` function was something that was added to both OSes after both forked. There is _no_ cross-pollination between the two projects. Oracle does their own thing and the only notice you get about anything is when a release comes out. illumos does what we feel is best for us -- while we strive to not break existing stuff, when adding new interfaces, while compatibility with Solaris may be a factor, it is far from the only (or most important) factor.

In this case, `getrandom()` was added after illumos forked from Solaris. I'm not sure which one got it first, but the illumos version was modeled after FreeBSD. It appears that Oracle changed the function signature in Solaris 11.4 so it was the same as the illumos and FreeBSD version (you can look at the section 2 man pages of Solaris 11.3 and Solaris 11.4).

The use of dlsym was to better support the existing `is_getrandom_available` test. Issuing unsupported syscalls on Solaris and illumos systems (as noted in the comments) generates a SYSSIG signal which is usually unhandled by a program (and thus usually crashes it). Checking for the libc function is less likely to cause problems.

Reply:
I don't know why you are even talking about Illumos. My question is entirely and exclusively about Solaris. For the purpose of this discussion, I am treating the two as completely independent OSes.

Reply:
Specifically, the question is, what is the return type for `getrandom` om Solaris? This PR made it `c_int`. But today in libc, it is `ssize_t`. Who is right? Or did Solaris change the return type of this function?

Reply:
>  It appears that Oracle changed the function signature in Solaris 11.4 

Tat;s our best running theory. It seems very strange to me that one can just change such a function signature but maybe that works for Oracle for some reason?

That hypothesis explains basically everything, except for https://www.gnu.org/software/gnulib/manual/html_node/getrandom.html which says
> This function has a different return type on some platforms: Solaris 11.4.